File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/quartz/controller/QuartzJobConfigurationController.java
package com.jmsoftware.maf.springcloudstarter.quartz.controller;

import com.jmsoftware.maf.common.bean.PageResponseBodyBean;
import com.jmsoftware.maf.common.bean.ResponseBodyBean;
import com.jmsoftware.maf.springcloudstarter.poi.AbstractExcelDataController;
import com.jmsoftware.maf.springcloudstarter.quartz.entity.CreateOrModifyQuartzJobConfigurationPayload;
import com.jmsoftware.maf.springcloudstarter.quartz.entity.GetQuartzJobConfigurationPageListItem;
import com.jmsoftware.maf.springcloudstarter.quartz.entity.GetQuartzJobConfigurationPageListPayload;
import com.jmsoftware.maf.springcloudstarter.quartz.entity.QuartzJobConfigurationExcel;
import com.jmsoftware.maf.springcloudstarter.quartz.service.QuartzJobConfigurationService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.time.Instant;
import java.util.List;

import static cn.hutool.core.text.CharSequenceUtil.format;

/**
 * <h1>QuartzJobConfigurationController</h1>
 * <p>
 * Change description here.
 *
 * @author Johnny Miller (鍾俊), email: johnnysviva@outlook.com, 9/26/21 11:58 PM
 **/
@Slf4j
@Validated
@RestController
@RequiredArgsConstructor
@RequestMapping("/quartz-job")
public class QuartzJobConfigurationController extends AbstractExcelDataController<QuartzJobConfigurationExcel> {
    private final QuartzJobConfigurationService service;

    @GetMapping("/quartz-job-configurations")
    public PageResponseBodyBean<GetQuartzJobConfigurationPageListItem> getPageList(
            @Valid GetQuartzJobConfigurationPageListPayload payload
    ) {
        return this.service.getPageList(payload);
    }

    @PostMapping("/quartz-job-configurations")
    public ResponseBodyBean<Long> create(
            @Valid @RequestBody CreateOrModifyQuartzJobConfigurationPayload payload
    ) {
        return ResponseBodyBean.ofSuccess(this.service.create(payload));
    }

    @PutMapping("/quartz-job-configurations/{id}")
    public ResponseBodyBean<Long> modify(
            @PathVariable Long id,
            @Valid @RequestBody CreateOrModifyQuartzJobConfigurationPayload payload
    ) {
        return ResponseBodyBean.ofSuccess(this.service.modify(id, payload));
    }

    @PatchMapping("/quartz-job-configurations/{id}/{property}")
    public ResponseBodyBean<Long> patch(
            @PathVariable Long id,
            @PathVariable String property,
            @RequestBody CreateOrModifyQuartzJobConfigurationPayload payload
    ) {
        return ResponseBodyBean.ofSuccess(this.service.patch(id, property, payload));
    }

    @PostMapping("/quartz-job-configurations/{id}/run-immediately")
    public ResponseBodyBean<Long> runImmediately(@PathVariable Long id) {
        return ResponseBodyBean.ofSuccess(this.service.runImmediately(id));
    }

    @DeleteMapping("/quartz-job-configurations/{id}/{group}")
    public ResponseBodyBean<Long> delete(@PathVariable Long id, @PathVariable String group) {
        return ResponseBodyBean.ofSuccess(this.service.delete(id, group));
    }

    @Override
    protected void onExceptionOccurred() {
        log.error("Exception occurred when uploading excel. Excel class: {}", QuartzJobConfigurationExcel.class);
        this.fileName.set(format("quartz-job-configuration-stat-{}.xlsx", Instant.now()));
    }

    @Override
    protected void validateBeforeAddToBeanList(List<QuartzJobConfigurationExcel> beanList,
                                               QuartzJobConfigurationExcel bean, int index)
            throws IllegalArgumentException {
        this.service.validateBeforeAddToBeanList(beanList, bean, index);
    }

    @Override
    protected void executeDatabaseOperation(List<QuartzJobConfigurationExcel> beanList) {
        this.service.save(beanList);
    }

    @Override
    protected String getTemplateFileName() {
        return QuartzJobConfigurationService.TEMPLATE_EXCEL;
    }

    @Override
    protected List<QuartzJobConfigurationExcel> getListForExporting() {
        return this.service.getListForExporting();
    }
}

File: spring-cloud-starter/src/main/kotlin/com/jmsoftware/maf/springcloudstarter/quartz/controller/QuartzJobConfigurationController.kt
package com.jmsoftware.maf.springcloudstarter.quartz.controller

import com.jmsoftware.maf.common.bean.PageResponseBodyBean
import com.jmsoftware.maf.common.bean.ResponseBodyBean
import com.jmsoftware.maf.common.util.LoggerDelegate
import com.jmsoftware.maf.springcloudstarter.poi.AbstractExcelDataController
import com.jmsoftware.maf.springcloudstarter.quartz.entity.CreateOrModifyQuartzJobConfigurationPayload
import com.jmsoftware.maf.springcloudstarter.quartz.entity.GetQuartzJobConfigurationPageListItem
import com.jmsoftware.maf.springcloudstarter.quartz.entity.GetQuartzJobConfigurationPageListPayload
import com.jmsoftware.maf.springcloudstarter.quartz.entity.QuartzJobConfigurationExcel
import com.jmsoftware.maf.springcloudstarter.quartz.service.QuartzJobConfigurationService
import jakarta.validation.Valid
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import java.time.Instant

/**
 * # QuartzJobConfigurationController
 *
 * Change description here.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/14/22 9:12 PM
 */
@Validated
@RestController
@RequestMapping("/quartz-job")
class QuartzJobConfigurationController(
    private val service: QuartzJobConfigurationService
) : AbstractExcelDataController<QuartzJobConfigurationExcel>() {
    companion object {
        private val log by LoggerDelegate()
    }

    @GetMapping("/configurations")
    fun getPageList(
        payload: @Valid GetQuartzJobConfigurationPageListPayload
    ): PageResponseBodyBean<GetQuartzJobConfigurationPageListItem> = service.getPageList(payload)

    @PostMapping("/configurations")
    fun create(
        @RequestBody payload: @Valid CreateOrModifyQuartzJobConfigurationPayload
    ): ResponseBodyBean<Long> = ResponseBodyBean.ofSuccess(service.create(payload))

    @PutMapping("/configurations/{id}")
    fun modify(
        @PathVariable id: Long,
        @Valid @RequestBody payload: CreateOrModifyQuartzJobConfigurationPayload
    ): ResponseBodyBean<Long> = ResponseBodyBean.ofSuccess(service.modify(id, payload))

    @PatchMapping("/configurations/{id}/{property}")
    fun patch(
        @PathVariable id: Long,
        @PathVariable property: String,
        @RequestBody payload: CreateOrModifyQuartzJobConfigurationPayload
    ): ResponseBodyBean<Long> = ResponseBodyBean.ofSuccess(service.patch(id, property, payload))

    @PostMapping("/configurations/{id}/run-immediately")
    fun runImmediately(@PathVariable id: Long): ResponseBodyBean<Long> =
        ResponseBodyBean.ofSuccess(service.runImmediately(id))

    @DeleteMapping("/configurations/{id}/{group}")
    fun delete(@PathVariable id: Long, @PathVariable group: String): ResponseBodyBean<Long> =
        ResponseBodyBean.ofSuccess(service.delete(id, group))

    override fun onExceptionOccurred() {
        log.error("Exception occurred when uploading excel. Excel class: `${QuartzJobConfigurationExcel::class.java}`")
        fileName.set("quartz-job-configuration-stat-${Instant.now()}.xlsx")
    }

    override fun validateBeforeAddToBeanList(
        beanList: List<QuartzJobConfigurationExcel>,
        bean: QuartzJobConfigurationExcel, index: Int
    ) = service.validateBeforeAddToBeanList(beanList, bean, index)

    override fun executeDatabaseOperation(beanList: List<QuartzJobConfigurationExcel>) = service.save(beanList)

    override fun getTemplateFileName(): String = QuartzJobConfigurationService.TEMPLATE_EXCEL

    override fun getListForExporting(): List<QuartzJobConfigurationExcel> = service.getListForExporting()
}
